android/app/src/main/java/com/expensify/chat/CustomNotificationProvider.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Check if we are showing a notification related to a reportID.
+     * If not, create a new NotificationCache so we can build a conversation notification
+     * as the messages come.
+     *
+     * @param reportID Report ID.
+     * @return Notification Cache.
+     */
+    private NotificationCache findOrCreateNotificationCache(int reportID) {
+        NotificationCache notificationCache = cache.get(reportID);
+
+        if (notificationCache == null) {
+            notificationCache = new NotificationCache();
+            cache.put(reportID, notificationCache);
+        }
+
+        return notificationCache;
+    }
+
+    /**
+     * Remove the notification data from the cache when the user dismisses the notification.
+     *
+     * @param message Push notification's message
+     */
+    public void onDismissNotification(PushMessage message) {
+        try {
+            JsonMap payload = JsonValue.parseString(message.getExtra(PAYLOAD_KEY)).optMap();
+            int reportID = payload.get("reportID").getInt(-1);
+
+            if (reportID == -1) {
+                return;
+            }
+
+            cache.remove(reportID);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to delete conversation cache");
+        }
+    }
+
+    @NonNull
+    private IconCompat fetchIcon(@NonNull String urlString, @DrawableRes int fallbackId) {
+        // TODO: Add disk LRU cache
+
+        URL parsedUrl = null;
+        try {
+            parsedUrl = urlString == null ? null : new URL(urlString);
+        } catch (MalformedURLException e) {
+            Log.e(TAG, "Failed to resolve URL " + urlString, e);
+        }
+
+        if (parsedUrl != null) {
+            WindowManager window = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
+            DisplayMetrics dm = new DisplayMetrics();
+            window.getDefaultDisplay().getMetrics(dm);
+
+            final int reqWidth = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, MAX_ICON_SIZE_DPS, dm);
+            final int reqHeight = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, MAX_ICON_SIZE_DPS, dm);
+
+            final URL url = parsedUrl;
+            Future<Bitmap> future = executorService.submit(() -> ImageUtils.fetchScaledBitmap(context, url, reqWidth, reqHeight));
+
+            try {
+                Bitmap bitmap = future.get(MAX_ICON_FETCH_WAIT_TIME_SECONDS, TimeUnit.SECONDS);
+                return IconCompat.createWithBitmap(bitmap);
+            } catch (InterruptedException e) {
+                Log.e(TAG,"Failed to fetch icon", e);
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                Log.e(TAG,"Failed to fetch icon", e);
+                future.cancel(true);
+            }
+        }
+
+        return fallbackId == 0 ? null : IconCompat.createWithResource(context, fallbackId);
+    }
+
+    private static class NotificationCache {
```

Comment:
Hmm, why does this need to be static? Is this to ensure only one instance of NotificationCache exists? 

Even as a static class, I would have thought that the cache would be lost when the app was killed? Unless `CustomNotificationProvider` has a lifecycle that extends beyond the life of the app? 😕

Reply:
There are multiple instances of NotificationCache, one for each conversation that gets shown to the user in a notification. We only remove the cache after the user dismisses or opens a notification.

From my understanding, the CustomNotificationProvider lives after the app gets killed as a background service. I don't know how reliable it is, though, as I have no experience with Android dev. If we shouldn't rely on it, maybe it would be interesting to store the notification cache data in SQLite or something similar?

Reply:
> There are multiple instances of NotificationCache, one for each conversation that gets shown to the user in a notification. 

In which case, I don't think the class should be static. 

>  If we shouldn't rely on it, maybe it would be interesting to store the notification cache data in SQLite or something similar?

That still leaves a dependency on a DB that will need to be run in the background. Ideally the data should be sent in the push notification payload, so we would send all unread messages. 

> CustomNotificationProvider lives after the app gets killed as a background service. 

This would be very odd if true. Would you mind sharing the documentation for this? Or any links you have to the code they gave us.

Reply:
It doesn't allow me to remove the static modifier of the class, and if I do it, we won't be able to do `new NotificationCache.Message(...);`

---

Yeah, I didn't explain correctly. This service gets launched when a notification arrives, but it will die after a while; as you say, we have no guarantee.

We can't send all data in the notification payload because of the payload size limits (~10KB). That's why it's better to cache the notification conversations on the device. I guess using the shared preferences could be enough for this use case instead of using a DB.

Also, having the conversation stored on the device will make it easier for us when we want to implement the quick reply from the notification.

Reply:
It seems that they are using making the cache class static, so that it's lifecycle extends that of each individual `ReactNotificationProvider` instance. It works I guess, but it's a bit hacky.

Reply:
> We can't send all data in the notification payload because of the payload size limits (~10KB)
But surely 10KB is fine for some simple objects? We don't need the full profile image but simply the URL, no?